pkg/pki/k8scsrpki/k8scsr_user.go

```diff
@@ -0,0 +1,266 @@
+// Copyright © 2021 Banzai Cloud
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package k8scsrpki
+
+import (
+	"context"
+	"crypto/x509"
+	"encoding/pem"
+	"fmt"
+
+	"emperror.dev/errors"
+
+	"github.com/banzaicloud/k8s-objectmatcher/patch"
+
+	"github.com/banzaicloud/kafka-operator/api/v1alpha1"
+	"github.com/banzaicloud/kafka-operator/pkg/errorfactory"
+	"github.com/banzaicloud/kafka-operator/pkg/util"
+
+	certutil "github.com/banzaicloud/kafka-operator/pkg/util/cert"
+	pkicommon "github.com/banzaicloud/kafka-operator/pkg/util/pki"
+
+	certsigningreqv1 "k8s.io/api/certificates/v1"
+	corev1 "k8s.io/api/core/v1"
+	apierrors "k8s.io/apimachinery/pkg/api/errors"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/runtime"
+	"k8s.io/apimachinery/pkg/types"
+	"sigs.k8s.io/controller-runtime/pkg/controller/controllerutil"
+)
+
+// ReconcileUserCertificate ensures and returns a user certificate - should be idempotent
+func (c *k8sCSR) ReconcileUserCertificate(
+	ctx context.Context, user *v1alpha1.KafkaUser, scheme *runtime.Scheme, _ string) (*pkicommon.UserCertificate, error) {
+	var clientKey []byte
+	var signingReq *certsigningreqv1.CertificateSigningRequest
+	secret := &corev1.Secret{}
+	err := c.client.Get(ctx, types.NamespacedName{Name: user.Spec.SecretName, Namespace: user.Namespace}, secret)
+	// Handle case when secret with private key is not found
+	if apierrors.IsNotFound(err) {
+		clientKey, err = certutil.GeneratePrivateKeyInPemFormat()
+		if err != nil {
+			return nil, err
+		}
+		secret = generateUserSecret(clientKey, user.Spec.SecretName, user.GetNamespace())
+		err = controllerutil.SetControllerReference(user, secret, scheme)
+		if err != nil {
+			return nil, err
+		}
+		if err = patch.DefaultAnnotator.SetLastAppliedAnnotation(secret); err != nil {
+			return nil, errors.WrapIf(err, "could not apply last state to annotation")
+		}
+		err = c.client.Create(ctx, secret)
+		if err != nil {
+			return nil, err
+		}
+		// Generate new SigningRequest resource
+		signingReq, err = c.generateAndCreateCSR(ctx, clientKey, user)
+		if err != nil {
+			return nil, err
+		}
+
+		if err = c.secretUpdateAnnotation(ctx, secret, signingReq.GetName()); err != nil {
+			return nil, err
+		}
+	} else if err != nil {
+		return nil, err
```

Comment:
Maybe add a little detail to the error?
```suggestion
		return nil, errors.WrapIfWithDetails(err, "failed to get user's secret from K8s", "secretName", user.Spec.SecretName, "namespace", user.GetNamespace())
```